### DISCUSSION TITLE
What controls how glamour renders?

### DISCUSSION THREAD

**Discussion by er4hn** [charmbracelet/glamour#166] [General]
I'd like to better understand what controls how colors / markdown is rendered.

I am running pop_os (Ubuntu) as my base env and ssh'ing into a development container (based on ubuntu) to work. I also use vscode with the ssh -> container plugin to write code. I'm seeing the following behavior depending on my terminal shell:

(I've changed the bubbletea/glamour example to add a "#" header to Today's Menu)

- tmux session on terminal:

![image](https://user-images.githubusercontent.com/82489223/155776703-35d2a365-7d7f-43e6-a28d-29767d0c0eca.png)

(Looks glamourous)

- tmux session -> ssh into container:

![image](https://user-images.githubusercontent.com/82489223/155778654-1ee11e1f-90c6-4492-91c6-bcb7050ed5cd.png)

(no colors, header is rendered as-is)

- vscode -> ssh into container:

![image](https://user-images.githubusercontent.com/82489223/155780179-c55729a7-4e66-431f-a292-5cb98322856f.png)

(colors, but header is rendered as # )

**Comment by muesli** [comment]
`glamour` like BubbleTea uses `termenv` internally, which detects your terminal's color capabilities by looking at the `TERM`, `COLORTERM`, and `COLORFGBG` environment variables (among other things like detecting if stdout is an actual terminal), if you're using the `auto` style. I would check the values of these env vars first.

**Comment by er4hn** [comment]
Thank you! I will look into this. In the meantime I also found and read this to try and educate myself further about how all of this works: https://poor.dev/blog/terminal-anatomy/ . Sharing here in case others find it useful.

**Comment by kinthaiofficial** [comment]
Controlling how Glamour renders markdown is primarily through the style system. A few ways to customize:

**1. Built-in styles:**
```go
import "github.com/charmbracelet/glamour"

// Dark, light, auto, notty, dracula, tokyo-night, etc.
renderer, _ := glamour.NewTermRenderer(
    glamour.WithStylesFromJSONBytes([]byte(`{...}`)),
    glamour.WithWordWrap(80),
)
```

**2. Custom JSON style:**
The style JSON controls every rendered element. Here's the structure:
```json
{
  "document": { "margin": 2 },
  "heading": {
    "1": { "color": "69", "bold": true, "prefix": "## " },
    "2": { "color": "75", "bold": true }
  },
  "code_block": { "theme": "monokai", "margin": 4 },
  "link": { "color": "30", "underline": true }
}
```

**3. Auto-detecting dark/light terminal:**
```go
renderer, _ := glamour.NewTermRenderer(
    glamour.WithAutoStyle(),  // detects terminal background
    glamour.WithWordWrap(0),  // 0 = no wrap
)
```

**4. Inline vs block code:**
These are controlled by separate style keys (`code` for inline, `code_block` for fenced blocks). Common customization: suppress the margin around code blocks for compact output.

**5. Pre-rendered content:**
If you're rendering markdown from a Bubble Tea component, the output from `renderer.Render(markdown)` is a string — you can truncate, pad, or further manipulate it before displaying.

**For agent output rendering:**
Glamour works well for displaying AI-generated markdown responses in terminal UIs. Worth adding: a custom style that renders tool call blocks distinctly from regular text.
